schema/impl.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/heapster/sinks/cache"
+)
+
+var lock sync.RWMutex
+
+func NewCluster() Cluster {
+	return newRealCluster()
+}
+
+func newRealCluster() *realCluster {
+	cinfo := ClusterInfo{
+		InfoType:   newInfoType(nil, nil),
+		Namespaces: make(map[string]*NamespaceInfo),
+		Nodes:      make(map[string]*NodeInfo),
+	}
+	cluster := &realCluster{
+		timestamp:   time.Time{},
+		ClusterInfo: cinfo,
+	}
+	return cluster
+}
+
+func (rc *realCluster) GetAllClusterData() (*ClusterInfo, time.Time, error) {
+	// Returns the entire ClusterInfo object
+	lock.RLock()
+	defer lock.RUnlock()
+	return &rc.ClusterInfo, rc.timestamp, nil
+}
+
+func (rc *realCluster) GetAllNodeData(hostname string) (*NodeInfo, time.Time, error) {
+	// Returns the corresponding NodeInfo object
+	lock.RLock()
+	defer lock.RUnlock()
+
+	var res *NodeInfo
+	var stamp time.Time
+	var err error
+
+	if val, ok := rc.Nodes[hostname]; ok {
+		res = val
+		stamp = rc.timestamp
+		err = nil
+	} else {
+		res = nil
```

Comment:
Why not `return nil, stamp, fmt.Errorf(...)` instead? 
Another style nit: The pattern that we usually follow is that of handling the failure case first and then moving on with the rest of the logic.